XMonad/Hooks/Modal.hs

```diff
@@ -89,35 +100,107 @@ import           XMonad.Util.Loggers
+-- while the latter is more geared towards how defining keys works by
+-- default in xmonad. Note that, by default, modes are exited with the
+-- Escape key. If one wishes to customise this, the 'modeWithExit'
+-- function should be used instead of 'mode' when defining a new mode.
+--
+-- The label of the active mode can be logged with 'logMode' to be
+-- displayed in a status bar, for example (For more information check
+-- "XMonad.Util.Loggers"). Some examples are included in [the provided
+-- modes](#g:ProvidedModes).
 
 -- }}}
 
 -- --< Types >-- {{{
 
--- | The mode type. Use 'mode' or 'mode'' to create modes.
-data Mode = Mode
-  { label     :: String
-  , boundKeys :: XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())
+-- | A class for the different kinds of ways the user could specify
+-- keys; each way should be an instance of this. This is used to give
+-- the user a little bit more freedom when specifying keys, while also
+-- allowing us to force a certain representation for internal use.
+class GetBoundKeys k where  -- type GetBoundKeys ∷ Type → Constraint
+  getBoundKeys :: k -> XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())
+
+-- | Internally, we always represent keys as the usual
+-- @XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())@ function.
+newtype KeysIntern
+  = KeysIntern (XConfig Layout -> M.Map (ButtonMask, KeySym) (X ()))
+
+instance GetBoundKeys KeysIntern where
+  getBoundKeys :: KeysIntern -> XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())
+  getBoundKeys = coerce
+
+-- | Create a 'Mode' holding the internal key representation; i.e., one
+-- that does not need to be parsed anymore.
+mkIntern :: Mode BoundKeys -> Mode KeysIntern
+mkIntern (Mode l bk) = Mode l $ KeysIntern (getBoundKeys bk)
+
+-- | Ways to supply a collection of bound keys to be used in a mode.
+-- Creation is done via the 'mkKeysFun' and 'mkKeysEz' functions.
+data BoundKeys
+  = KeysEZ  ![(String, X ())]
+  | KeysFun !KeysIntern
+
+instance GetBoundKeys BoundKeys where
+  getBoundKeys :: BoundKeys -> XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())
+  getBoundKeys bk cnf = case bk of
+    KeysEZ  k  -> mkKeymap cnf k
+    KeysFun ki -> getBoundKeys ki cnf
+
+-- | From a function that returns keybindings, generate a key
+-- representation.
+--
+-- >>> mkKeysFun $ const $ M.fromList [((noModMask, xK_h), xmessage "Hello, World!")]
+mkKeysFun :: (XConfig Layout -> M.Map (ButtonMask, KeySym) (X ())) -> BoundKeys
+mkKeysFun = KeysFun . KeysIntern
+
+-- | From a list of 'XMonad.Util.EZConfig'-style bindings, generate a
+-- key representation.
+--
+-- >>> mkKeysEz [("h", xmessage "Hello, world!")]
+mkKeysEz :: [(String, X ())] -> BoundKeys
+mkKeysEz = KeysEZ
+
+-- | The mode type. Use 'mode' or 'modeWithExit' to create modes.
+data Mode k = Mode
+  { label     :: !String
+  , boundKeys :: !k
   }
 
+-- | A convenient alias for specifying new modes.
+type NewMode = Mode BoundKeys
+
```

Comment:
Wouldn't 
```haskell
data Mode = Mode
  { label     :: !String
  , boundKeys :: !BoundKeys
  }
```
do the work? I don't see where we're setting the type variable to anything else

Reply:
> I don't see where we're setting the type variable to anything else

For the extensible config/state we're using

``` haskell
newtype ModeConfig = MC [Mode KeysIntern] deriving Semigroup
```

and

``` haskell
newtype CurrentMode = CurrentMode
  { currentMode :: Maybe (Mode KeysIntern)
  }
```

This is the main reason for both the type variable there, as well as the `GetBoundKeys` class.  If we just had, for example, `newtype ModeConfig = MC [Mode]` (with a baked in `BoundKeys`) then we couldn't—at the type level—ensure that the internal representation of `XConfig Layout -> Map (ButtonMask, KeySym) (X ())` is used in the state.  We'd have to do it via some note or contract of "please do this", which seemed a bit dodgy to me; especially since the solution isn't much more complicated.

Reply:
Right, but since we're not exposing the types anyway, do we really need `BoundKeys`? What I actually meant was

```haskell
data Mode = Mode
  { label     :: !String
  , boundKeys :: !KeysIntern
  }
```
with the following interface:

```haskell
sayHelloMode :: NewMode
sayHelloMode = mode "Hello" $ mkKeysEz
  [ ("h", xmessage "Hello, World!")
  , ("g", xmessage "Goodbye, World!")
  ]


sayHelloMode :: NewMode
sayHelloMode = mode "Hello" $
  const (M.fromList [ ((noModMask, xK_h), xmessage "Hello, World!")
                    , ((noModMask, xK_g), xmessage "Goodbye, World!")
                    ])
```

and the new signatures (eventually renaming `KeysIntern`):
```
mkKeysEz :: [(String, X ())] -> KeysIntern
modeWithExit :: String -> String -> KeysIntern -> NewMode
```

Reply:
Oh!  Why yes, this make everything a lot easier :sweat_smile: